apps/hash-graph/lib/authorization/src/zanzibar.rs

```diff
@@ -0,0 +1,264 @@
+#[derive(Debug)]
+pub struct Tuple<R, A, S> {
+    pub resource: R,
+    pub affiliation: A,
+    pub subject: S,
+}
+
+impl<R, A, S> fmt::Display for Tuple<R, A, S>
+where
+    R: Resource,
+    A: Affiliation<R>,
+    S: Subject,
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            fmt,
+            "{}:{}#{}@{}:{}",
+            self.resource.namespace(),
+            self.resource.id(),
+            self.affiliation,
+            self.subject.resource().namespace(),
+            self.subject.resource().id()
+        )?;
+        if let Some(affiliation) = self.subject.affiliation() {
+            write!(fmt, "#{affiliation}")?;
+        }
+        Ok(())
+    }
+}
+
+/// An untyped [`Tuple`] that is generic over the [`Resource`], [`Affiliation`], and [`Subject`].
+///
+/// This is useful for when the [`Resource`], [`Affiliation`], and [`Subject`] types are not known
+/// at compile-time, e.g. when parsing a [`Tuple`] from a string.
+pub type StringTuple = Tuple<
+    GenericResource<String, String>,
+    GenericAffiliation<String>,
+    GenericSubject<GenericResource<String, String>, GenericAffiliation<String>>,
+>;
+
+/// Provide causality metadata between Write and Check requests.
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(transparent)]
+pub struct Zookie<'a>(Cow<'a, str>);
+
+/// Specifies the desired consistency level on a per-request basis.
+///
+/// This allows for the API consumers dynamically trade-off less fresh data for more performance
+/// when possible.
+#[derive(Debug)]
+pub enum Consistency<'z> {
+    /// Attempts to minimize the latency of the API call, using whatever caches are available.
+    ///
+    /// > ## Warning
+    /// >
+    /// > If used exclusively, this can lead to a window of time where the New Enemy Problem can
+    /// > occur.
+    MinimalLatency,
+    /// Ensures that all data used for computing the response is at least as fresh as the
+    /// point-in-time specified in the [`Zookie`].
+    ///
+    /// If newer information is available, it will be used.
+    AtLeastAsFresh(Zookie<'z>),
+    /// Ensures that all data used for computing the response is that found at the exact
+    /// point-in-time specified in the [`Zookie`].
+    ///
+    /// If the snapshot is not available, an error will be raised.
+    AtExactSnapshot(Zookie<'z>),
+    /// Ensure that all data used is fully consistent with the latest data available within the
+    /// SpiceDB datastore.
+    ///
+    /// Note that the snapshot used will be loaded at the beginning of the API call, and that new
+    /// data written after the API starts executing will be ignored.
+    ///
+    /// > ## Warning
+    /// >
+    /// > Use of `FullyConsistent` means little caching will be available, which means performance
+    /// > will suffer. Only use if a [`Zookie`] is not available or absolutely latest information
+    /// > is required.
+    FullyConsistent,
+}
```

Comment:
Didn't miss this file, @TimDiekmann and I talked about this in detail offline already, so no comments :P

Reply:
Did you know that you can comment on a file? 🙂 
<img width="200" alt="image" src="https://github.com/hashintel/hash/assets/21277928/9b905ff5-8f44-435f-b7b8-817f64ba1341">

Reply:
d'oh, didn't think about that when (I just woke up when I wrote the review^^) :P